src/compiler/graph/link_nodes.rs

```diff
@@ -0,0 +1,239 @@
+        Some(match get_facing_macro!(0, X, diff) {
+            Some(x) => x,
+            None => match get_facing_macro!(1, Y, diff) {
+                Some(y) => y,
+                None => get_facing_macro!(2, Z, diff)?,
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use test_case::test_case;
+
+    use super::*;
+    use crate::{types::block::Facing, utils::test::*};
+
+    #[test_case((0, 0, 0), (0, 0, 1), Facing::NegativeZ ; "facing neg z")]
+    #[test_case((0, 0, 1), (0, 0, 0), Facing::PositiveZ ; "facing pos z")]
+    #[test_case((0, 1, 0), (0, 0, 0), Facing::PositiveY ; "facing pos y")]
+    #[test_case((0, 0, 0), (0, 1, 0), Facing::NegativeY ; "facing neg y")]
+    #[test_case((1, 0, 0), (0, 0, 0), Facing::PositiveX ; "facing pos x")]
+    #[test_case((0, 0, 0), (1, 0, 0), Facing::NegativeX ; "facing neg x")]
+    fn test_util_get_facing(pos1: Position, pos2: Position, facing: Facing) {
+        let pos1 = vec![pos1.0, pos1.1, pos1.2];
+        let pos2 = vec![pos2.0, pos2.1, pos2.2];
+
+        assert_eq!(utils::get_facing(pos1, pos2).unwrap(), facing);
+    }
+
+    // Dust
+    #[test_case(&make_block!(kind: Kind::Component(Component::Dust), pos: (0, 0, 1), facing: vec![Facing::NegativeZ]), Some(Link::new_weak()) ; "dust pointing into block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Dust), pos: (0, 0, 1), facing: vec![Facing::PositiveX]), None ; "dust not pointing into block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Dust), pos: (0, 1, 0)), Some(Link::new_weak()) ; "dust ontop of block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Dust), pos: (0, -1, 0)), None ; "dust under block")]
+    // Repeater
+    #[test_case(&make_block!(kind: Kind::Component(Component::new_repeater()), pos: (0, 0, 1), facing: vec![Facing::NegativeZ]), Some(Link::StrongPower) ; "repeater pointing into block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::new_repeater()), pos: (0, 1, 0)), None ; "repeater ontop of block")]
+    // Lever
+    #[test_case(&make_block!(kind: Kind::Component(Component::Lever { on: false }), pos: (1, 0, 0), facing: vec![Facing::NegativeX]), Some(Link::StrongPower) ; "lever on block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Lever { on: false }), pos: (1, 0, 0), facing: vec![Facing::PositiveX]), None ; "lever not on block")]
+    // Tourch
+    // Shouldn't matter where it is facing
+    #[test_case(&make_block!(kind: Kind::Component(Component::Tourch { lit: false }), pos: (0, -1, 0)),  Some(Link::StrongPower) ; "tourch under block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Tourch { lit: false }), pos: (0, 1, 0), facing: vec![Facing::NegativeY]),  None ; "tourch ontop block")]
+    #[test_case(&make_block!(kind: Kind::Component(Component::Tourch { lit: false }), pos: (0, 0, 1), facing: vec![Facing::PositiveZ]),  None ; "tourch next to block")]

```

Comment:
Probably best that we don't do it this way as it will spam the tests with lots of tests